pennylane/devices/default_clifford.py

```diff
@@ -466,50 +478,256 @@ def simulate(
+
+        return meas_results[0] if len(meas_results) == 1 else tuple(meas_results)
 
     @staticmethod
     def pl_to_stim(op):
         """Convert PennyLane operation to a Stim operation"""
         try:
             stim_op = _GATE_OPERATIONS[op.name]
         except KeyError as e:
-            raise DeviceError(
+            raise qml.DeviceError(
                 f"Operator {op} not supported on default.clifford and does not provide a decomposition."
             ) from e
         return stim_op, op.wires
 
-    def measure(self, circuit, tableau_simulator, global_phase, stim):
-        """Given a circuit, compute and return the measurement results."""
+    @staticmethod
+    def _from_hamiltonian_or_tensor(op):
+        """Convert old PennyLane observable to an arithmetic operator"""
+        if isinstance(op, qml.Hamiltonian):  # pragma: no cover
+            return qml.dot(*op.terms())
+        if isinstance(op, qml.operation.Tensor):
+            return qml.prod(*op.obs)
+        return op
+
+    # pylint:disable=too-many-return-statements
+    def _convert_op_to_linear_comb(self, meas_obs, coeffs, paulis):
+        """Convert a PennyLane observable to a linear combination of stim Pauli terms"""
+
+        # Case for simple Pauli terms
+        if isinstance(meas_obs, (qml.Identity, qml.PauliZ, qml.PauliX, qml.PauliY)):
+            coeffs.append(1.0)
+            paulis.append((_GATE_OPERATIONS[meas_obs.name], meas_obs.wires))
+            return coeffs, paulis
+
+        # Case for simple Pauli tensor
+        if isinstance(meas_obs, qml.operation.Tensor):
+            coeffs.append(1.0)
+            paulis.append(
+                ("".join([_GATE_OPERATIONS[name] for name in meas_obs.name]), meas_obs.wires)
+            )
+            return coeffs, paulis
+
+        # Case for a Hamiltonian
+        if isinstance(meas_obs, qml.Hamiltonian):
+            cof, obs = meas_obs.terms()
+            coeffs.extend(cof)
+            for ob in obs:
+                expec = (
+                    _GATE_OPERATIONS[ob.name]
+                    if not isinstance(ob.name, list)
+                    else "".join([_GATE_OPERATIONS[o] for o in ob.name])
+                )
+                paulis.append((expec, ob.wires))
+            return coeffs, paulis
+
+        # Case for Sum
+        if isinstance(meas_obs, Sum):
+            for op in meas_obs:
+                coeffs, paulis = self._convert_op_to_linear_comb(op, coeffs, paulis)
+            return coeffs, paulis
+
+        # Case for higher arithmetic depth for prod-type observables
+        if meas_obs.arithmetic_depth > 1 and isinstance(meas_obs, (Prod, SProd)):
+            check_op_math = qml.operation.active_new_opmath()
+            qml.operation.enable_new_opmath()
+            meas_obs_simp = meas_obs.simplify()
+            if not check_op_math:
+                qml.operation.disable_new_opmath()
```

Comment:
I _think_ there should be a more elegant way to do this. I don't know what it is off the top of my head, but I'd like to confirm with an expert. I'm also curious if my recent bugfixes circumvent the need for this. Can we try removing it and re-running tests? If it's still a hassle, then nbd

Reply:
There might be a more elegant way, but this was the most straightforward one that came to me. This specific `if` branch deals with the case where I have an object built using nested `Prod` and `SProd`.

Reply:
I was able to reduce the logic for the old operators, but I would be keeping the logic for new ones. :)